README.md
# DbSequel

### mysql
```
/**
 * application: labti-server
 * module: models/mysql/labti/index.js
 */

const { join, sep } = require('path');
const dbsequel = require('sicinfo-dbsequel');

const dialect = 'mysql'
const schema = 'labti';

module.exports = options => {

  dialect in options || (options[dialect] = {});

  if (!(schema in options[dialect])) {

    const {
      database, username, password, 
      host = 'localhost', 
      port = '3306', 
      pool = {}, 
      define = {},
      prefixTableName = `${schema}_`
    } = require(join(process.env.HOME, options.dbconfig))[dialect][schema];

    'max' in pool || (pool.max = 5);
    'min' in pool || (pool.min = 0);
    'idle' in pool || (pool.idle = 10000);
    'acquire' in pool || (pool.acquire = 30000);

    'timestamps' in define || (define.timestamps = false);
    'underscored' in define || (define.underscored = false);
    'freezeTableName' in define || (define.freezeTableName = true);
    
    options[dialect][schema] = dbsequel({
      database, username, password, prefixTableName,
      'options': { pool, define, dialect, host, port },
      'dirmodels': __dirname
    })

  }

  return options[dialect][schema];
};
```
### postgres
```
/**
 * application: labti-server
 * module: models/postgres/citsmart/index.js
 */

const { join, sep } = require('path');
const dbsequel = require('sicinfo-dbsequel');

const dialect = 'postgres';
const schema = 'citsmart';

module.exports = options => {

  dialect in options || (options[dialect] = {});

  if (!(schema in options[dialect])) {

    const {
      database, username, password, 
      host = 'localhost', 
      port = '5432', 
      pool = {}
    } = require(join(process.env.HOME, options.dbconfig))[dialect][schema];

    'max' in pool || (pool.max = 5);
    'min' in pool || (pool.min = 0);
    'idle' in pool || (pool.idle = 10000);
    'acquire' in pool || (pool.acquire = 30000);

    options[dialect][schema] = dbsequel({
      database, username, password,
      'options': { pool, dialect, host, port },
      'dirmodels': __dirname
    })

  }

  return options[dialect][schema];
};
```

### apllication/grupo
```
/***
 * application: labti
 * module: application/grupo.js
 * 
 * Powered [0.0.1] by Moreira in 2017-12-06
 * 
 */
'use strict';

const { join } = require('path');

module.exports = options => new Promise((resolve, reject) => {
  
  const model = require(join(options.dirname, 'models', 'mysql', 'labti'))(options)
  .then(ff => ff('Grupo'));
  
  resolve(opts => {

    return model.then(({fetch}) => fetch(opts));

  })

});
```


### schema para grupo
```
/**
 * application: labti
 * module: models/mysql/labti/xx-Grupo
 */

module.exports.schema = (Datatypes, done) => {

  return done({
    'nome': { 'type': Datatypes.STRING },
    'id': Datatypes.PRIMARYKEY
  })

};
```

### application/index
```
/**
 * application labti
 * module: application/index.js
*/

const { join } = require('path');

module.exports = ({ dirname, version }) => {

  const { config } = require(join(dirname, 'package.json'));
  Object.assign(config, { dirname, version, 'mysql': {}, 'postgres': {} });

  return (name, promise) => {

    return ({ query, params, method }, res) => {

      promise || (promise = require(`./${name}`)(config))

      return promise
        .then(next => {

          const opts = { 'where': query, method };

          Object.keys(params)
            .filter(k => undefined !== params[k] && isNaN(k))
            .forEach(k => opts[k] = params[k]);

          return next.call(this, opts)
        })
        .then(resp => {
          
          if (resp.count) {
            res.append('X-total-count', resp.count);
          }

          return res.json(resp.rows);
        })
        .catch(error => {
          return res.status(404).json({error})
        });
    }
  };

};
```

### index
```
/**
 * application: labti
 * module: index.js
 * 
 * Updated: 
 * [0.0.10] Moreira
 * - inclui verison
 * 
 * [0.0.1] Moreira 2017-11-08
 *
 ************************************************/

'use strict';

const express = require('express');
const { json } = require('body-parser');
const { join } = require('path');

const { version } = require(join(__dirname, '..', 'package.json'));
const application = require('./application')({'dirname': __dirname, version});

const app = module.exports = express();

app.get('/version', (req, res) => res.json({ version }))

app.get('/etiqueta(/:id)?', application('etiqueta'));

app.get('/grupo(/:id)?', application('grupo'));

app.get('/solicitacaoservico/:idsolicitacaoservico', application('solicitacaoservico'));

```

CHANGELOG.md
## DbSequel

### [2.1.1] 2020-03-11 by moreira
- bugfix: fetchDocument()

### [2.1.0] 2020-03-04 by moreira
- correçoes diversas

### [2.0.0] 2019-07-15 by moreira
- altera estruturação dos objectos
- inclui classes para extender sequelize.Model 

### [1.0.0] 2019-06-26 by Moreira 
- inclui dependencia de sicinfo-idcreate no package.json
- correções diversas

### [1.0.0] 2019-06-26 by Moreira 
- inclui dependencia de sicinfo-idcreate no package.json
- correções diversas

### [0.0.8] 2018-06-28
#### melhoria
- altera schema do arquivos model
- teste do método save, update e delete

### [0.0.6] 2018-06-22
#### altera
- altera assinatura do modulo para function no lugar de =>

### [0.0.5] 2018-05-29
- optimiza
- corrige dbsequel

### [0.0.4] 2018-05-29
- altera asmore-id -> sicinfo-idcreate

### [0.0.4] 2018-05-29
- correções

### [0.0.2] 2018-05-29

### [0.0.1] 2018-05-25


libs.1/dbsequel.copia.js
/**
 * application: sicinfo-dbsequel
 * module: lib/dbsequel.js
 * 
 * Updated [1.0.0] by Moreira in 2019-06-22
 * 
 * 
 * Powered [0.0.1] by Moreira in 2017-05-25
 **/

const 
  Sequelize = require('sequelize'),
  { newId } = require('sicinfo-idcreate'),
  { Op } = Sequelize,
  Datatypes = { Sequelize, newId };

[
  'INTEGER', 
  'TINYINT', 
  'REAL', 
  'DATE', 
  'ENUM', 
  'STRING', 
  'TEXT', 
  'VIRTUAL'
].forEach(key => Datatypes[key] = Sequelize[key]);


Datatypes.SECUNDARYKEY = {
  'type': Datatypes.STRING,
  'primaryKey': true
};

Datatypes.PRIMARYKEY = {
  'type': Datatypes.STRING,
  'primaryKey': true,
  'defaultValue': () => newId()
};

Datatypes.FOREIGNKEY = {
  'type': Datatypes.STRING,
  'allowNull': false,
  'onDelete': 'RESTRICT',
  'onUpdate': 'RESTRICT'
};

Datatypes.foreignKey = (references, options = {}) => {
  options['references'] = references;
  Object.keys(Datatypes.FOREIGNKEY).forEach(elem => {
    elem in options || (options[elem] = Datatypes.FOREIGNKEY[elem]);
  });
  return options;
};

const parseQueryValues = one => resp => {
  if (Array.isArray(resp)) { resp = { 'rows': resp } }
  if (resp.rows.length == 0) { resp.rows = one ? {} : [] }
  else if (one) { resp.rows = resp.rows[0] }
  else {
    const head = Object.keys(resp.rows[0]);
    resp.rows = [head].concat(resp.rows.map(row => head.map(key => row[key])));
  }
  return resp;
};

const parseFindValues = one => resp => {
  if (Array.isArray(resp)) { resp = { 'rows': resp } }
  if (resp.rows.length == 0) { resp.rows = one ? {} : [] }
  else {

    const incl = {}, { rows } = resp, data = [[]];
    let pos = 0;

    Object.keys(rows[0].dataValues).some(key => {
      const val = rows[0].dataValues[key];
      if (val && 'object' === typeof (val) && 'dataValues' in val) {
        incl[key] = true;
        Object.keys(val.dataValues).some(key => { data[0][pos++] = key });
      }
      else data[0][pos++] = key;
    });

    rows.forEach((row, ind) => {
      data[ind + 1] = [];
      pos = 0;
      Object.keys(row.dataValues).forEach(key => {
        const val = row.dataValues[key];
        if (incl[key]) Object.keys(val.dataValues).forEach(key => {
          data[ind + 1][pos++] = val.dataValues[key];
        });
        else {
          data[ind + 1][pos++] = val;
        }
      });
    });

    if (one) {
      resp.rows = {};
      data[0].forEach((key, ind) => resp.rows[key] = data[1][ind]);
    }
    else resp.rows = data;
  }

  return resp;
};

// const failFind = done => err => done(err.name && (
//   `${err.name} - ${err.original && err.original.code || ''}`    
// ) || err);
const _failDone = err => {

  console.log(err);

  const msg = err.original && (
    `${err.name} - ${err.original.code || err.original}`
  ) /*|| err.message*/ || err;

  return msg;
};

const _query = (sequelize, sqlString, options = {}) => {
  'type' in options || (options.type = Sequelize.QueryTypes.SELECT);
  return sequelize.query(options.one ? `${sqlString} LIMIT 1` : sqlString, options)
    .then(parseQueryValues(options.one))
    .catch(_failDone);
};

const _fetch = (model, options = {}) => {
  // força relacao INNER JOIN
  'include' in options && options['include'].forEach((include, index) => {
    'required' in include || (options['include'][index]['required'] = true);
  });

  return (model.primaryKeyAttributes.every(
    primaryKeyAttribute => primaryKeyAttribute in options
  ) ? _findById : _findAll)(model, options);
};

const _findAll = (model, options = {}, _options = {}) => {

  if ('where' in options === false) {
    options.where = {};
  }

  _options.byId || Object.keys(options.where)
    .filter(key => options.where[key].split('').some(a => '%' === a))
    .map(key => { options.where[key] = { [Op.like]: options.where[key] } });

  return model[`find${options.count ? 'AndCount' : ''}All`](options)
    .then(parseFindValues(_options.one))
    .catch(_failDone);
};

const _findOne = (model, options = {}) => {
  options.limit = 1;
  return _findAll(model, options, { 'one': true });
};

const _findById = (model, options = {}) => {
  
  options.where = {};
  model.primaryKeyAttributes
    .filter(key => key in options)
    .forEach(key => { options.where[key] = options[key] });

  return _findAll(model, options, { 'byId': true, 'one': true });
};

const _update = (model, options) => {

  const where = options.where || {};
  model.primaryKeyAttributes.map(key => {
    key in where || (where[key] = options[key]);
  });

  const data = {};
  Object.keys(options.data)
    .filter(key => !(key in where))
    .some(key => data[key] = options.data[key]);

  return model.update(data, { where }).then(rows => ({ rows }));
};

const _delete = (model, options) => {

  if (!model.primaryKeyAttributes.every(key => key in options)) {
    return new Promise((a, b) => b('erro'));
  }

  const where = options.where || {};
  model.primaryKeyAttributes.map(key => {
    key in where || (where[key] = options[key]);
  });

  return model.destroy({ where }).then(rows => ({ rows }));
};

const _save = (model, options) => {

  if (model.primaryKeyAttributes.every(key => key in options)) {
    return _update(model, options);
  }

  return model.create(options.data);
};

/**
 * @param {*} opts 
 */
module.exports = function ({ database, username, password, options = {}, dirmodels, prefixTableName = '' }) {

  if ('operatorsAliases' in options === false) {
    options.operatorsAliases = Op;
  }

  const 
    sequelize = new Sequelize(database, username, password, options), 
    relations = {}, 
    files = {};

  return new Promise((resolve, reject) => sequelize.authenticate()
    .then(() => require('fs').readdir(dirmodels, (err, args) => {

      // console.log(dirmodels);

      if (err) return reject({ err });

      // filtrar somente arquivos com nomes no padrão
      args.filter(arg => /\d\d-[A-Z]/.test(arg.slice(0, 4)) && arg.endsWith('.js'))
        .some(_filename => {

          // ler e o arquivos para Sequelize
          const name = _filename.split('.')[0].slice(3);
          const modelFile = files[name] = require(require('path').join(dirmodels, _filename));

          // 0 - estrutura
          // 1 - opções
          // 2 - relações
          modelFile.schema(Datatypes, (schemas) => {

            if ('schema' === Object.keys(schemas)[0]) {
              schemas = (obj => (obj[name] = schemas, obj))({});
            }

            Object.keys(schemas).forEach(name => {
              const { schema, options = {}, associations = {} } = schemas[name];

              if (!('tableName' in options)) {
                options.tableName = `${prefixTableName}${name.toLowerCase()}`;
              }

              Object.keys(schema)
                .filter(attr => 'object' === typeof (schema[attr]) && 'references' in schema[attr])
                .some(foreignKey => {
                  const 
                    belongsTo = associations.belongsTo || (associations.belongsTo = {}),
                    { references } = schema[foreignKey],
                    options = belongsTo[references] || (belongsTo[references] = {});
                  options['foreignKey'] = foreignKey;
                });

              if (associations) {
                relations[name] = associations;
              }

              return sequelize.define(name, schema, options);
            });
          });
        });

      Object.keys(relations).forEach(filename => {
        ['belongsTo', 'hasOne', 'hasMany', 'belongsToMany']
          .filter(association => association in relations[filename])
          .forEach(association => {
            Object.keys(relations[filename][association]).forEach(target => {
              let options = relations[filename][association][target];
              if ('string' === typeof (options)) options = { 'foreignKey': options };
              sequelize.models[filename][association](sequelize.models[target], options);
            });
          });
      });

      // arg -> nome do model
      // quando não tiver model o paramentro arg deve ser null
      // e será retornado o objeto sequelize
      return resolve(modelName => {

        // console.log(modelName);

        if ('string' === typeof (modelName)) {
          const _call = _back => options => {

            // console.log(modelName);
            // console.log(options);

            return _back(sequelize.models[modelName], options);
          };
          
          return {
            'fetch': _call(_fetch),
            'findAll': _call(_findAll),
            'findOne': _call(_findOne),
            'findById': _call(_findById),
            'save': _call(_save),
            'update': _call(_update),
            'delete': _call(_delete),
            'drop': _call(_delete),
            'models': sequelize.models
          };
        }

        else return {
          'query': (sqlString, options) => {
            return _query(sequelize, sqlString, options);
          },
          'models': sequelize.models
        };
      });
    }))
  );

};
